In [1]:
import requests
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

Define a function named get_lower_and_upper_bounds that has two arguments. The first argument is a pandas Series. The second argument is the multiplier, which should have a default argument of 1.5.

In [2]:
def get_lower_and_upper_bounds(pd_series, multiplier = 1.5):
    """
    It creates the lower and the upper bounds of a series with a default multiplier of 1.5
    """
    #
    q1 = pd_series.quantile(0.25)
    q3 = pd_series.quantile(0.75)
    # calculate iqr
    iqr = q3 - q1
    # calculate upper and lower limits for the whiskers
    inner_lower_fence = q1 - (multiplier * iqr)
    inner_upper_fence = q3 + (multiplier * iqr)
    return inner_lower_fence, inner_upper_fence 

--------------------------
# 1 
Using lemonade.csv dataset and focusing on continuous variables:

In [3]:
# set a variable named URL where we place the HTTPS as a string 
url = 'https://gist.githubusercontent.com/ryanorsinger/19bc7eccd6279661bd13307026628ace/raw/e4b5d6787015a4782f96cad6d1d62a8bdbac54c7/lemonade.csv'

In [4]:
# Reed data as a apnda dataferame especifing the delimiter in order to create tabular data
df = pd.read_csv(url,sep=",") # use sep="," for coma separation. 
df.head()


Date        Day  Temperature  Rainfall  Flyers  Price  Sales
0  1/1/17     Sunday         27.0      2.00      15    0.5     10
1  1/2/17     Monday         28.9      1.33      15    0.5     13
2  1/3/17    Tuesday         34.5      1.33      27    0.5     15
3  1/4/17  Wednesday         44.1      1.05      28    0.5     17
4  1/5/17   Thursday         42.4      1.00      33    0.5     18

In [5]:
# Change the mame of the columns to lower case
df.columns = df.columns.str.lower()

In [6]:
df.describe()

temperature    rainfall      flyers  price       sales
count   365.000000  365.000000  365.000000  365.0  365.000000
mean     61.224658    0.825973   40.104110    0.5   27.865753
std      18.085892    0.273890   13.786445    0.0   30.948132
min      15.100000    0.400000  -38.000000    0.5    7.000000
25%      49.700000    0.650000   31.000000    0.5   20.000000
50%      61.100000    0.740000   39.000000    0.5   25.000000
75%      71.700000    0.910000   49.000000    0.5   30.000000
max     212.000000    2.500000   80.000000    0.5  534.000000

# 1.1
Use the IQR Range Rule and the upper and lower bounds to identify the lower outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these lower outliers make sense? Which outliers should be kept?

In [7]:
# set upper and lower bound fot df.temperature
temp_lower, temp_upper = get_lower_and_upper_bounds(df.temperature)

In [8]:
lower_temp_outliers = df[temp_lower > df.temperature]
lower_temp_outliers

date     day  temperature  rainfall  flyers  price  sales
364  12/31/17  Sunday         15.1       2.5       9    0.5      7

In [9]:
# set upper and lower bound fot df.reinfall
rainfall_lower, rainfall_upper = get_lower_and_upper_bounds(df.rainfall)

In [10]:
# Get the lower outliers for rainfall 
lower_rainfall_outliers = df[rainfall_lower > df.rainfall]
lower_rainfall_outliers

Empty DataFrame
Columns: [date, day, temperature, rainfall, flyers, price, sales]
Index: []

In [11]:
# set upper and lower bound fot df.flyers
flyers_lower, flyers_upper = get_lower_and_upper_bounds(df.flyers)
# Get the lower outliers for flyers 
lower_flyers_outliers = df[flyers_lower > df.flyers]
lower_flyers_outliers

date      day  temperature  rainfall  flyers  price  sales
324  11/21/17  Tuesday         47.0      0.95     -38    0.5     20

In [12]:
# set upper and lower bound fot df.price
price_lower, price_upper = get_lower_and_upper_bounds(df.price)
# Get the lower outliers for price 
lower_price_outliers = df[price_lower > df.price]
lower_price_outliers

Empty DataFrame
Columns: [date, day, temperature, rainfall, flyers, price, sales]
Index: []

In [13]:
# set upper and lower bound fot df.
sales_lower, sales_upper = get_lower_and_upper_bounds(df.sales)
# Get the lower outliers for price 
lower_sales_outliers = df[sales_lower > df.sales]
lower_sales_outliers

Empty DataFrame
Columns: [date, day, temperature, rainfall, flyers, price, sales]
Index: []


Yes, this ouliers make sence. In fact, the only two columns return outliers and are the most minimum values in their range.

# 1.2
Use the IQR Range Rule and the upper and upper bounds to identify the upper outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these upper outliers make sense? Which outliers should be kept?

In [14]:
upper_temp_outliers = df[temp_upper < df.temperature]
upper_temp_outliers

date       day  temperature  rainfall  flyers  price  sales
41  2/11/17  Saturday        212.0      0.91      35    0.5     21

In [15]:
upper_rainfall_outliers = df[rainfall_upper < df.rainfall]
upper_rainfall_outliers

date        day  temperature  rainfall  flyers  price  sales
0      1/1/17     Sunday         27.0      2.00      15    0.5     10
1      1/2/17     Monday         28.9      1.33      15    0.5     13
2      1/3/17    Tuesday         34.5      1.33      27    0.5     15
5      1/6/17     Friday         25.3      1.54      23    0.5     11
6      1/7/17   Saturday         32.9      1.54      19    0.5     13
10    1/11/17  Wednesday         32.6      1.54      23    0.5     12
11    1/12/17   Thursday         38.2      1.33      16    0.5     14
12    1/13/17     Friday         37.5      1.33      19    0.5     15
15    1/16/17     Monday         30.6      1.67      24    0.5     12
16    1/17/17    Tuesday         32.2      1.43      26    0.5     14
19    1/20/17     Friday         31.6      1.43      20    0.5     12
23    1/24/17    Tuesday         28.6      1.54      20    0.5     12
27    1/28/17   Saturday         34.9      1.33      15    0.5     13
28    1/29/17     Sunday         35.2      1.33      27    0.5     14
337   12/4/17     Monday         34.9      1.54      16    0.5     13
338   12/5/17    Tuesday         22.0      1.82      11    0.5     10
342   12/9/17   Saturday         31.2      1.43      19    0.5     14
343  12/10/17     Sunday         31.3      1.82      15    0.5     11
345  12/12/17    Tuesday         33.5      1.33      22    0.5     15
346  12/13/17  Wednesday         32.2      1.43      26    0.5     14
347  12/14/17   Thursday         31.9      1.54      24    0.5     13
350  12/17/17     Sunday         32.2      1.33      16    0.5     14
351  12/18/17     Monday         30.9      1.43      27    0.5     13
354  12/21/17   Thursday         40.5      1.33      23    0.5     15
355  12/22/17     Friday         30.9      1.54      17    0.5     13
359  12/26/17    Tuesday         28.9      1.43      23    0.5     13
363  12/30/17   Saturday         30.9      1.43      22    0.5     13
364  12/31/17     Sunday         15.1      2.50       9    0.5      7

In [19]:
df.rainfall.describe()

count    365.000000
mean       0.825973
std        0.273890
min        0.400000
25%        0.650000
50%        0.740000
75%        0.910000
max        2.500000
Name: rainfall, dtype: float64

There are a lot of outliers in this range. I would keep all records within three standard deviations, meaning anything below 1.55

In [16]:
upper_flyers_outliers = df[temp_upper < df.flyers]
upper_flyers_outliers

Empty DataFrame
Columns: [date, day, temperature, rainfall, flyers, price, sales]
Index: []

In [17]:
upper_price_outliers = df[temp_upper < df.price]
upper_price_outliers

Empty DataFrame
Columns: [date, day, temperature, rainfall, flyers, price, sales]
Index: []

In [18]:
upper_sales_outliers = df[temp_upper < df.sales]
upper_sales_outliers

date       day  temperature  rainfall  flyers  price  sales
181  7/1/17  Saturday        102.9      0.47      59    0.5    143
182  7/2/17    Sunday         93.4      0.51      68    0.5    158
183  7/3/17    Monday         81.5      0.54      68    0.5    235
184  7/4/17   Tuesday         84.2      0.59      49    0.5    534

In [22]:
df.sales.describe()

count    365.000000
mean      27.865753
std       30.948132
min        7.000000
25%       20.000000
50%       25.000000
75%       30.000000
max      534.000000
Name: sales, dtype: float64

In [25]:
#calculate 3 standar diviations for sales
df.sales.describe()[5]

25.0

In [26]:
std = df.sales.describe()[2]
mid = df.sales.describe()[5]

In [27]:
# Upper limit for outliers 
mid + std + std + std 

117.84439595691894

Eliminate outliers in the sales colum taht are above 117.844

# 1.3
Using the multiplier of 3, IQR Range Rule, and the lower bounds, identify the outliers below the lower bound in each colum of lemonade.csv. Do these lower outliers make sense? Which outliers should be kept?

In [40]:
# create a list with the name of the vaibles we want to find the lower and the upper limits to 
new_cols = ['temperature', 'rainfall', 'flyers', 'price', 'sales']

In [50]:
# This for loop attachess the upper and the lower limmits by using get_lower_and_upper_bounds function 
#we create a dictionay that holds all the values and attaches and record of the data frame
multiplier_of_3 = {}
for col in new_cols:
    lower, upper = get_lower_and_upper_bounds(df[col], multiplier = 3)
    multiplier_of_3[col] = {'upper' : upper, 'lower' : lower}
    multiplier_of_3[col]['df'] = df[(df[col] > upper) | (df[col] < lower)]

In [54]:
multiplier_of_3['temperature']['df']

date       day  temperature  rainfall  flyers  price  sales
41  2/11/17  Saturday        212.0      0.91      35    0.5     21

In [55]:
list(multiplier_of_3.keys())

['temperature', 'rainfall', 'flyers', 'price', 'sales']

In [56]:
multiplier_of_3

{'temperature': {'upper': 137.7,
  'lower': -16.299999999999997,
  'df':        date       day  temperature  rainfall  flyers  price  sales
  41  2/11/17  Saturday        212.0      0.91      35    0.5     21},
 'rainfall': {'upper': 1.69,
  'lower': -0.13,
  'df':          date      day  temperature  rainfall  flyers  price  sales
  0      1/1/17   Sunday         27.0      2.00      15    0.5     10
  338   12/5/17  Tuesday         22.0      1.82      11    0.5     10
  343  12/10/17   Sunday         31.3      1.82      15    0.5     11
  364  12/31/17   Sunday         15.1      2.50       9    0.5      7},
 'flyers': {'upper': 103.0,
  'lower': -23.0,
  'df':          date      day  temperature  rainfall  flyers  price  sales
  324  11/21/17  Tuesday         47.0      0.95     -38    0.5     20},
 'price': {'upper': 0.5,
  'lower': 0.5,
  'df': Empty DataFrame
  Columns: [date, day, temperature, rainfall, flyers, price, sales]
  Index: []},
 'sales': {'upper': 60.0,
  'lower': -10.0,

In [60]:
for _.lower in multiplier_of_3:
    print(lower)

AttributeError: 'dict' object has no attribute 'lower'

In [63]:
for col in multiplier_of_3:
    multiplier_of_3[col]['lower']

# 1.4
Using the multiplier of 3, IQR Range Rule, and the upper bounds, identify the outliers above the upper_bound in each colum of lemonade.csv. Do these upper outliers make sense? Which outliers should be kept?